internal/auth/ldap/repository_managed_group.go

```diff
@@ -0,0 +1,261 @@
+package ldap
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/oplog"
+	"github.com/hashicorp/go-dbw"
+)
+
+// CreateManagedGroup inserts an ManagedGroup, mg, into the repository and
+// returns a new ManagedGroup containing its PublicId. mg is not changed. mg
+// must contain a valid AuthMethodId. mg must not contain a PublicId. The
+// PublicId is generated and assigned by this method. All options are ignored.
+//
+// Both mg.Name and mg.Description are optional. If mg.Name is set, it must be
+// unique within mg.AuthMethodId.
+func (r *Repository) CreateManagedGroup(ctx context.Context, scopeId string, mg *ManagedGroup, _ ...Option) (*ManagedGroup, error) {
+	const op = "ldap.(Repository).CreateManagedGroup"
+	switch {
+	case mg == nil:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing managed group")
+	case mg.ManagedGroup == nil:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing embedded managed group")
+	case mg.AuthMethodId == "":
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing auth method id")
+	case len(mg.GroupNames) == 0:
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing group names")
+	case mg.PublicId != "":
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "public id must be empty")
+	case scopeId == "":
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "missing scope id")
+	}
+
+	mg = mg.clone()
+
+	id, err := newManagedGroupId(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	mg.PublicId = id
+
+	oplogWrapper, err := r.kms.GetWrapper(ctx, scopeId, kms.KeyPurposeOplog)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg("unable to get oplog wrapper"), errors.WithCode(errors.Encrypt))
+	}
+
+	oplogMetadata, err := mg.oplog(ctx, oplog.OpType_OP_TYPE_CREATE, scopeId)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg("unable to generate managed group oplog metadata"))
+	}
+
+	var newManagedGroup *ManagedGroup
+	_, err = r.writer.DoTx(ctx, db.StdRetryCnt, db.ExpBackoff{},
+		func(_ db.Reader, w db.Writer) error {
+			newManagedGroup = mg.clone()
```

Comment:
the first clone is needed to ensure that mg is not changed (part of the contract) before we add the public id. This clone is needed to ensure that retried transactions have an changed copy to work on (since create could change it).